File: ui_playwright_bdd/scripts/agents/jira-ticket-agent.js
// ============================================
// Jira Ticket Agent — Création automatique de tickets
// ============================================
// Lit les résultats allure-results, analyse les échecs avec Groq,
// et crée automatiquement des tickets Bug dans Jira.
// Vérifie les doublons avant création.
//
// Usage:
//   npm run agent:ticket              → tickets depuis allure-results
//   npm run agent:ticket -- --dry-run → simulation sans créer
//   npm run agent:ticket -- --type=Story → crée des Stories (défaut: Bug)
//
// Output:
//   docs/JIRA_TICKETS_CREATED.md     — rapport des tickets créés
// ============================================

require('dotenv').config();
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

const llm  = require('./llm');
const jira = require('./jira-fetcher');
const fs   = require('fs');
const path = require('path');

const projectRoot = path.resolve(__dirname, '../../');
const resultsDir  = path.join(projectRoot, 'allure-results');
const docsDir     = path.join(projectRoot, 'docs');

// ── Arguments CLI ─────────────────────────────────────────────────────────────
const DRY_RUN     = process.argv.includes('--dry-run');
const ISSUE_TYPE  = (process.argv.find(a => a.startsWith('--type=')) || '--type=Bug').split('=')[1];
const ISSUE_TYPE_IDS = { Bug: '10005', Story: '10004', Task: '10003' };
const ISSUE_TYPE_ID  = ISSUE_TYPE_IDS[ISSUE_TYPE] || '10005';

// ── Lit les échecs allure-results ─────────────────────────────────────────────
function readFailures() {
  if (!fs.existsSync(resultsDir)) return [];
  return fs.readdirSync(resultsDir)
    .filter(f => f.endsWith('-result.json'))
    .map(f => {
      try { return JSON.parse(fs.readFileSync(path.join(resultsDir, f), 'utf-8')); }
      catch { return null; }
    })
    .filter(r => r && (r.status === 'failed' || r.status === 'broken'));
}

// ── Récupère les tickets existants pour éviter les doublons ──────────────────
async function fetchExistingTitles() {
  try {
    const stories = await jira.fetchStories();
    // Récupère aussi les bugs (type différent)
    const { body } = await require('https').get ? fetchBugs() : { body: { issues: [] } };
    return new Set(stories.map(s => s.summary.toLowerCase()));
  } catch { return new Set(); }
}

async function fetchBugs() {
  const https = require('https');
  const token = Buffer.from(`${process.env.JIRA_EMAIL}:${process.env.JIRA_TOKEN}`).toString('base64');
  const base  = (process.env.JIRA_BASE_URL || '').replace(/\/$/, '');
  const project = process.env.JIRA_PROJECT || 'SCRUM';
  const jql = encodeURIComponent(`project = ${project} AND issuetype = Bug ORDER BY created DESC`);
  return new Promise((resolve, reject) => {
    https.get(`${base}/rest/api/3/search/jql?jql=${jql}&maxResults=100&fields=summary`, {
      headers: { Authorization: `Basic ${token}`, Accept: 'application/json' }
    }, res => {
      let d = ''; res.on('data', c => d += c);
      res.on('end', () => {
        try { resolve(JSON.parse(d)); } catch { resolve({ issues: [] }); }
      });
    }).on('error', () => resolve({ issues: [] }));
  });
}

async function fetchAllExistingTitles() {
  const titles = new Set();
  try {
    const stories = await jira.fetchStories();
    stories.forEach(s => titles.add(s.summary.toLowerCase()));
    const bugsResp = await fetchBugs();
    (bugsResp.issues || []).forEach(i => titles.add((i.fields?.summary || '').toLowerCase()));
  } catch { /* ignore */ }
  return titles;
}

// ── Demande à Groq de structurer les tickets ──────────────────────────────────
async function generateTickets(failures) {
  const failureText = failures.map((f, i) => {
    const error = f.statusDetails?.message || 'No error message';
    const trace = (f.statusDetails?.trace || '').split('\n').slice(0, 8).join('\n');
    const steps = (f.steps || []).map(s => `  [${s.status}] ${s.name}`).slice(0, 6).join('\n');
    return `### Échec ${i + 1}: ${f.name}\n**Status:** ${f.status}\n**Erreur:** ${error}\n**Steps:**\n${steps}\n**Trace:**\n${trace}`;
  }).join('\n\n---\n\n');

  const messages = [
    {
      role: 'system',
      content: `Tu es un expert QA. Pour chaque échec de test, génère un ticket Jira structuré.
Réponds UNIQUEMENT avec un tableau JSON valide, sans texte autour.
Format de chaque ticket :
{
  "summary": "titre court et clair (max 100 chars)",
  "description": "description complète avec contexte, étapes pour reproduire, résultat attendu vs obtenu",
  "priority": "Highest|High|Medium|Low",
  "labels": ["regression", "automated-test"],
  "test_name": "nom du test concerné"
}`
    },
    {
      role: 'user',
      content: `Génère un ticket Jira de type ${ISSUE_TYPE} pour chaque échec ci-dessous.\n\n${failureText}`
    }
  ];

  const resp = await llm.chat(messages);
  const content = resp.message?.content || '[]';

  // Extrait le JSON de la réponse
  const match = content.match(/\[[\s\S]*\]/);
  if (!match) {
    console.warn('LLM response non-JSON, génération manuelle...');
    return failures.map(f => ({
      summary: `[Bug] ${f.name} — ${f.status}`,
      description: `**Test échoué:** ${f.name}\n**Status:** ${f.status}\n**Erreur:** ${f.statusDetails?.message || 'N/A'}\n**Trace:**\n${f.statusDetails?.trace?.slice(0, 800) || 'N/A'}`,
      priority: 'High',
      labels: ['regression', 'automated-test'],
      test_name: f.name
    }));
  }

  try { return JSON.parse(match[0]); }
  catch { return []; }
}

// ── Crée un ticket Bug dans Jira ──────────────────────────────────────────────
async function createTicket(ticket) {
  const https = require('https');
  const token  = Buffer.from(`${process.env.JIRA_EMAIL}:${process.env.JIRA_TOKEN}`).toString('base64');
  const base   = (process.env.JIRA_BASE_URL || '').replace(/\/$/, '');
  const project = process.env.JIRA_PROJECT || 'SCRUM';

  const body = JSON.stringify({
    fields: {
      project:     { key: project },
      issuetype:   { id: ISSUE_TYPE_ID },
      summary:     ticket.summary,
      description: {
        type: 'doc', version: 1,
        content: [{ type: 'paragraph', content: [{ type: 'text', text: ticket.description }] }]
      },
      priority: { name: ticket.priority || 'Medium' },
      labels:   ticket.labels || ['automated-test']
    }
  });

  return new Promise((resolve, reject) => {
    const url = new URL(`${base}/rest/api/3/issue`);
    const req = https.request({
      hostname: url.hostname,
      path: url.pathname,
      method: 'POST',
      headers: {
        Authorization: `Basic ${token}`,
        'Content-Type': 'application/json',
        Accept: 'application/json',
        'Content-Length': Buffer.byteLength(body)
      }
    }, res => {
      let d = ''; res.on('data', c => d += c);
      res.on('end', () => {
        try {
          const r = JSON.parse(d);
          resolve({ status: res.statusCode, key: r.key, error: r.errors });
        } catch { resolve({ status: res.statusCode, error: d }); }
      });
    });
    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function run() {
  await llm.assertRunning();

  const me = await jira.assertJira();
  console.log(`✅ Jira connecté — ${me.displayName}`);
  console.log(`\n=== JIRA TICKET AGENT  [${llm.MODEL}] ===`);
  console.log(`Projet: ${process.env.JIRA_PROJECT}  |  Type: ${ISSUE_TYPE}${DRY_RUN ? '  |  DRY-RUN' : ''}\n`);

  // Lit les échecs
  const failures = readFailures();
  console.log(`📋 Échecs trouvés dans allure-results: ${failures.length}`);

  if (!failures.length) {
    console.log('✅ Aucun échec — aucun ticket à créer.');

    // Propose des tickets depuis le RAG si pas d'échecs
    const bugAnalysis = path.join(docsDir, 'BUG_ANALYSIS.md');
    if (fs.existsSync(bugAnalysis)) {
      console.log('\n💡 BUG_ANALYSIS.md trouvé — vérification des bugs documentés...');
    }
    return;
  }

  // Vérifie les doublons
  console.log('🔍 Vérification des doublons dans Jira...');
  const existingTitles = await fetchAllExistingTitles();
  console.log(`   ${existingTitles.size} ticket(s) existant(s) indexé(s)\n`);

  // Génère les tickets avec Groq
  console.log('🤖 Génération des tickets avec LLM...');
  const rawTickets = await generateTickets(failures);

  // Déduplique les tickets générés (même summary)
  const seen = new Set();
  const tickets = rawTickets.filter(t => {
    const key = t.summary.toLowerCase().trim();
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
  console.log(`   ${rawTickets.length} généré(s) → ${tickets.length} unique(s) après déduplication\n`);

  // Crée les tickets dans Jira
  const created = [];
  const skipped = [];

  for (const ticket of tickets) {
    const isDuplicate = existingTitles.has(ticket.summary.toLowerCase());

    if (isDuplicate) {
      console.log(`⏭️  Doublon ignoré: ${ticket.summary.slice(0, 60)}`);
      skipped.push(ticket);
      continue;
    }

    if (DRY_RUN) {
      console.log(`[DRY-RUN] Créerait: ${ticket.summary.slice(0, 70)}`);
      created.push({ key: 'DRY-RUN', summary: ticket.summary });
      continue;
    }

    const result = await createTicket(ticket);
    if (result.status === 201) {
      const url = `${process.env.JIRA_BASE_URL}/browse/${result.key}`;
      console.log(`✅ ${result.key} créé — ${ticket.summary.slice(0, 60)}`);
      console.log(`   ${url}`);
      created.push({ key: result.key, summary: ticket.summary, url, priority: ticket.priority });

      // Ajoute le ticket créé aux doublons pour les itérations suivantes
      existingTitles.add(ticket.summary.toLowerCase());
    } else {
      console.error(`❌ Échec création: ${JSON.stringify(result.error)}`);
    }
  }

  // ── Rapport final ────────────────────────────────────────────────────────────
  const ts = new Date().toISOString();
  const report = [
    '# Jira Tickets Created — Rapport',
    '',
    `_${ts} — Projet: ${process.env.JIRA_PROJECT} — Type: ${ISSUE_TYPE} — ${llm.MODEL}_`,
    '',
    `## Résumé`,
    `- Échecs analysés : ${failures.length}`,
    `- Tickets créés   : ${created.length}`,
    `- Doublons ignorés: ${skipped.length}`,
    '',
    '## Tickets créés',
    '',
    created.length
      ? created.map(t => `- [${t.key}](${t.url || '#'}) [${t.priority || 'Medium'}] — ${t.summary}`).join('\n')
      : '_Aucun ticket créé._',
    '',
    '## Doublons ignorés',
    '',
    skipped.length
      ? skipped.map(t => `- ${t.summary}`).join('\n')
      : '_Aucun._',
    ''
  ].join('\n');

  fs.mkdirSync(docsDir, { recursive: true });
  fs.writeFileSync(path.join(docsDir, 'JIRA_TICKETS_CREATED.md'), report, 'utf-8');

  console.log(`\n📄 Rapport: docs/JIRA_TICKETS_CREATED.md`);
  console.log(`🎟️  ${created.length} ticket(s) créé(s)  |  ${skipped.length} doublon(s) ignoré(s)`);
}

run().catch(err => { console.error('Jira ticket agent error:', err.message || err); process.exit(1); });

File: mobile/android-appium-testng/agents/jira-ticket-agent.py
# ============================================================
# Jira Ticket Agent — Mobile (Appium / Android)
# ============================================================
# Lit les résultats Allure (target/allure-results/), génère
# des tickets Bug Jira structurés pour chaque échec mobile.
#
# Contexte ticket mobile :
#   - Device : Android Emulator (UiAutomator2)
#   - Framework : Appium 9.2.2 + TestNG 7.10.2
#   - App : QAcart-To-Do.apk
#   - Steps to reproduce : login → navigate → assert
#
# Usage:
#   python agents/jira-ticket-agent.py             → crée les tickets
#   python agents/jira-ticket-agent.py --dry-run   → affiche sans créer
#   python agents/jira-ticket-agent.py --type=Bug  → type d'issue (défaut: Bug)
# ============================================================

import sys, os, json, glob
sys.stdout.reconfigure(encoding="utf-8", errors="replace")
sys.path.insert(0, os.path.dirname(__file__))

from jira_fetcher_agent import JiraClient
import llm

DRY_RUN    = "--dry-run" in sys.argv
ISSUE_TYPE = next((a.split("=")[1] for a in sys.argv if a.startswith("--type=")), "Bug")
RESULTS_DIR = os.path.join(os.path.dirname(__file__), "..", "target", "allure-results")

TICKET_SCHEMA = {
    "summary":     "string — résumé du bug en 1 ligne (max 100 chars), préfixé [MOBILE]",
    "description": "string — description complète avec : Device, Framework, Steps to reproduce, Expected vs Actual, Error message",
    "priority":    "Low | Medium | High | Highest",
    "labels":      ["mobile", "appium", "android"]
}

G = "\033[32m"; R = "\033[31m"; Y = "\033[33m"; C = "\033[36m"; W = "\033[1m"; E = "\033[0m"


def load_failures() -> list:
    failures = []
    pattern  = os.path.join(RESULTS_DIR, "*-result.json")
    for f in glob.glob(pattern):
        try:
            with open(f, encoding="utf-8") as fh:
                d = json.load(fh)
            if d.get("status") not in ("failed", "broken"):
                continue

            labels     = d.get("labels", [])
            test_class = next((lb["value"] for lb in labels if lb["name"] == "testClass"), "?")
            short_cls  = test_class.split(".")[-1] if "." in test_class else test_class
            groups     = [lb["value"] for lb in labels if lb["name"] == "tag"]
            detail     = d.get("statusDetails") or {}

            failures.append({
                "name":       d.get("name", "?"),
                "test_class": short_cls,
                "status":     d.get("status"),
                "groups":     groups,
                "message":    detail.get("message", "")[:400],
                "trace":      detail.get("trace",   "")[:500],
            })
        except Exception:
            pass
    return failures


def generate_tickets(failures: list) -> list:
    if not failures:
        return []

    failures_json = json.dumps([{
        "test_class": f["test_class"],
        "name":       f["name"],
        "groups":     f["groups"],
        "status":     f["status"],
        "message":    f["message"][:200],
        "trace":      f["trace"][:200],
    } for f in failures], ensure_ascii=False, indent=2)

    prompt = f"""Tu es un expert QA Mobile (Appium/Android). Pour chaque échec de test ci-dessous,
génère un ticket Bug Jira structuré.

Contexte du projet :
- Device : Android Emulator (UiAutomator2)
- Framework : Appium 9.2.2 + TestNG 7.10.2 (Java 17)
- App sous test : QAcart-To-Do.apk
- Locators : AppiumBy.androidUIAutomator() / AppiumBy.className()

Réponds UNIQUEMENT avec un JSON valide : liste d'objets, chacun avec :
  summary     : "[MOBILE] Résumé court du bug"
  description : "Device: Android Emulator\\nFramework: Appium 9.2.2\\nTest: <classe.méthode>\\n\\nSteps to reproduce:\\n1. Lancer l'app\\n2. ...\\n\\nExpected: ...\\nActual: ...\\n\\nError:\\n<message d'erreur>"
  priority    : "Low | Medium | High | Highest"
  labels      : ["mobile", "appium", "android", <group>]

Règles priorité mobile :
  Highest → app crash, session Appium perdue, smoke test échoué
  High    → élément non trouvé sur écran principal
  Medium  → timeout, test de régression
  Low     → false positive, test en quarantaine

Échecs à analyser :
{failures_json}"""

    raw   = llm.chat([{"role": "user", "content": prompt}])
    start = raw.find("[")
    end   = raw.rfind("]") + 1
    if start == -1:
        return []

    try:
        tickets = json.loads(raw[start:end])
    except Exception:
        return []

    seen    = set()
    deduped = []
    for t in tickets:
        key = t.get("summary","").lower().strip()
        if key not in seen:
            seen.add(key)
            deduped.append(t)
    return deduped


def run():
    print(f"\n{W}=== JIRA TICKET AGENT MOBILE [{llm.MODEL}] ==={E}")
    if DRY_RUN:
        print(f"   {Y}MODE DRY-RUN — aucun ticket ne sera créé{E}\n")

    failures = load_failures()
    print(f"  {C}{len(failures)} échec(s) trouvé(s) dans target/allure-results/{E}")
    if not failures:
        print(f"  {G}✓ Aucun échec — aucun ticket à créer.{E}")
        return

    for f in failures:
        group_str = ", ".join(f["groups"]) or "aucun"
        print(f"  {R}✗{E} [{f['test_class']}] {f['name']} ({group_str})")

    print(f"\n  {C}Génération des tickets via LLM ({ISSUE_TYPE})...{E}")
    tickets = generate_tickets(failures)
    print(f"  {G}{len(tickets)} ticket(s) généré(s){E}")

    if DRY_RUN:
        print()
        for t in tickets:
            prio = t.get("priority", "?")
            prio_color = R if prio in ("Highest","High") else Y if prio == "Medium" else G
            print(f"  [{prio_color}{prio}{E}] {t.get('summary','?')}")
            labels = t.get("labels", [])
            if labels:
                print(f"    Labels : {', '.join(labels)}")
        return

    jira     = JiraClient()
    existing = {s["fields"]["summary"].lower() for s in jira.get_stories()}

    created = 0
    for t in tickets:
        if t.get("summary","").lower() in existing:
            print(f"  {Y}⏭  Doublon ignoré : {t.get('summary','')}{E}")
            continue

        extra_labels = list(set(["mobile", "appium", "android"] + t.get("labels", [])))
        issue = jira.create_story(
            summary     = t.get("summary", "Mobile Test Failure"),
            description = t.get("description", ""),
            priority    = t.get("priority", "Medium"),
        )
        key = issue.get("key", "?")
        print(f"  {G}✓ Créé : {key} — {t.get('summary','')}{E}")
        created += 1

    print(f"\n  {G}✓ {created} ticket(s) créé(s) dans Jira.{E}")


if __name__ == "__main__":
    run()
